feat: add `math/base/special/sincosd`

anandkaranubc opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

- Adds the JS and C implementation of the function to compute the sincos of `x`, where `x` is in degrees.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: add JS implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: add tests and its fixtures

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: add C implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: add benchmarks, examples, docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
[Reference implementation](https://github.com/JuliaLang/julia/blob/d63adeda50d5b8c440a7c29c9a65357a98fe927f/base/special/trig.jl#L1242-L1249)

anandkaranubc commented:
Also, the `cephes` and built-in math benchmarks have been removed because there is no implementation of `sincosd` in those libraries.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+# sincosd
+
+> Simultaneously compute the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of a number.
```
```suggestion
> Simultaneously compute the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an angle measured in degrees.
```

gunjjoshi replied:
@anandkaranubc Let's use an `angle` instead of a `number` here and everywhere else in this package, similar to what we have in [`sind`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/sind/README.md?plain=1#L23).

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/sincosd/lib/main.js:
```diff
@@ -0,0 +1,57 @@
+/**
+* Simultaneously computes the sine and cosine of a number.
+*
+* @param {number} x - input value (in degrees)
```
@gunjjoshi, I took the `sincos` implementation as a reference for this PR.

That is the reason I have the details about the input value here.

https://github.com/stdlib-js/stdlib/blob/aec0e8848ed94f8df559aec221a457ff3e27f8f2/lib/node_modules/%40stdlib/math/base/special/sincos/lib/main.js#L31

anandkaranubc replied:
Considering this, would you say the change is still needed?

gunjjoshi replied:
You're right to think so. But, in general, `sine` and `cosine` of a `number` doesn't make much sense, as compared to `sine` and `cosine` of an `angle`. In terms of the details of the input values, I think it would be much better to follow the packages added more recently, such as https://github.com/stdlib-js/stdlib/blob/aec0e8848ed94f8df559aec221a457ff3e27f8f2/lib/node_modules/%40stdlib/math/base/special/sind/README.md?plain=1#L23 and https://github.com/stdlib-js/stdlib/blob/aec0e8848ed94f8df559aec221a457ff3e27f8f2/lib/node_modules/%40stdlib/math/base/special/cosd/README.md?plain=1#L23, as compared to a relatively older one, i.e., https://github.com/stdlib-js/stdlib/blob/aec0e8848ed94f8df559aec221a457ff3e27f8f2/lib/node_modules/%40stdlib/math/base/special/sincos/README.md?plain=1#L23

anandkaranubc replied:
Sure! Thanks for the feedback. On it.

anandkaranubc committed:
Merge branch 'stdlib-js:develop' into feat/sincosd

anandkaranubc committed:
docs: clarify that inputs to sind and cosd are in degrees

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/docs/repl.txt:
```diff
@@ -0,0 +1,59 @@
+    [ NaN, NaN ]
+
+
+{{alias}}( x, out, stride, offset )
```
```suggestion
{{alias}}.assign( x, out, stride, offset )
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/lib/index.js:
```diff
@@ -0,0 +1,67 @@
+*
+* var out = new Float64Array( 2 );
+*
+* var v = sincosd( out, 0.0 );
```
```suggestion
* var v = sincosd.assign( 0.0, out, 1, 0 );
```

anandkaranubc replied:
On it!

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/src/main.c:
```diff
@@ -0,0 +1,40 @@
+*/
+void stdlib_base_sincosd( const double x, double* sine, double* cosine ) {
+	*sine = stdlib_base_sind( x );
+	*cosine = stdlib_base_cosd( x );
```
```suggestion
	*cosine = stdlib_base_cosd( x );
	return;
```

gunjjoshi replied:
Missing return statement.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/test/test.assign.js:
```diff
@@ -0,0 +1,383 @@
+	z = [ 0.0, 0.0 ];
+	v = sincosd( NaN, z, 1, 0 );
+	t.equal( v, z, 'returns output array' );
+	t.equal( isnan( v[0] ), true, 'returns NaN' );
```
```suggestion
	t.equal( isnan( v[0] ), true, 'returns expected value' );
```

gunjjoshi replied:
@anandkaranubc Can you change this everywhere else in this file and in other test files too, if applicable?

anandkaranubc replied:
Sure! On it right away! Thanks.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/package.json:
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "@stdlib/math/base/special/sincosd",
+  "version": "0.0.0",
+  "description": "Simultaneously compute the sine and cosine an angle measured in degrees.",
```
```suggestion
  "description": "Simultaneously compute the sine and cosine of an angle measured in degrees.",
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+# sincosd
+
+> Simultaneously compute the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] an angle measured in degrees.
```
```suggestion
> Simultaneously compute the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an angle measured in degrees.
```

anandkaranubc replied:
Oh sorry, that's my fault. I will change it in all the files.

gunjjoshi replied:
Thanks!

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+#### sincosd( x )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of a `number` (in degrees).
```
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+#### sincosd.assign( x, out, stride, offset )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of a `number` (in degrees) and assigns results to a provided output array.
```
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees and assigns results to a provided output array.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+#### stdlib_base_sincosd( x, &sine, &cosine )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of a `number` (in degrees).
```
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees.
```

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="links">
+
+<!-- <related-links> -->
+
+[@stdlib/math/base/special/cosd]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/cos
+
+[@stdlib/math/base/special/sind]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/sin
+
+<!-- </related-links> -->
```
```suggestion
<section class="links">

[@stdlib/math/base/special/cosd]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/cosd

[@stdlib/math/base/special/sind]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/sind

<!-- <related-links> -->

<!-- </related-links> -->
```

anandkaranubc committed:
chore: add missing 'the' in comment

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+# sincosd
+
+> Simultaneously compute the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees.
```
```suggestion
> Simultaneously compute the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an angle measured in degrees.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+#### sincosd( x )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees.
```
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an angle measured in degrees.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+#### sincosd.assign( x, out, stride, offset )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees and assigns results to a provided output array.
```
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an angle measured in degrees and assigns results to a provided output array.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+#### stdlib_base_sincosd( x, &sine, &cosine )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees.
```
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an angle measured in degrees.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/README.md:
```diff
@@ -0,0 +1,204 @@
+
+#### sincosd.assign( x, out, stride, offset )
+
+Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an `angle` measured in degrees and assigns the results to a provided output array.
```
```suggestion
Simultaneously computes the [sine][@stdlib/math/base/special/sind] and [cosine][@stdlib/math/base/special/cosd] of an angle measured in degrees and assigns the results to a provided output array.
```

anandkaranubc replied:
😔

gunjjoshi committed:
Apply suggestions from code review

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/sincosd/docs/repl.txt:
```diff
@@ -0,0 +1,60 @@
+
+
+{{alias}}.assign( x, out, stride, offset )
+    Simultaneously computes the sine and cosine of
+    an angle measured in degrees and assigns the
+    results to a provided output array.
```
```suggestion
    Simultaneously computes the sine and cosine of an angle measured in degrees
    and assigns the results to a provided output array.
```

gunjjoshi committed:
Update lib/node_modules/@stdlib/math/base/special/sincosd/docs/repl.txt

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi commented:
Thanks for working on this, @anandkaranubc!

anandkaranubc commented:
> Thanks for working on this, @anandkaranubc!

@gunjjoshi, big thanks to you as well for your review and feedback! Appreciate it! 🙌
